# xQTLImp-dist
## Introduction
xQTLImp-dist is a variant model of xQTLImp. xQTLImp-dist considers the biological context of xQTLs and assigns larger weights to the unknown variants that are closer to a trait. The input data needed is same as the xQTLImp, and their usages are similar except we added new parameter --delta which is a hyper-parameter of activation function(see Methods in our manuscript).

### xQTLImp parametersï¼š
The parameters can be specified by short options (e.g. -h) or long options (e.g. --help), which have same effects. Please see the parameters and explanations as follows:
```bash
xQTLImp
-h, --help           null           # null, to display this usage.
-x, --xQTL           file_path      # string, the file path of xQTL summary statistics.
-m, --molecule       file_path      # string, the file path of molecule annotation file.
-v, --VCF            folder_path    # string, the folder path of genome reference panel, such as 1000G VCF files.
-o, --output         folder_path    # string, the folder path of output results. 
-c, --chr            chromosome     # int, specify which chromosome will be imputed.
-t, --num_threads    num_threads    # int, number of threads, 1 in default.
-s, --sort           TRUE/FALSE     # boolean, sort the xQTL summary statistics by chromosome, molecular_ID, and variant_pos in increasing order prior imputation (required), FALSE in default.
-e, --exclude        chr:start-end  # int:int-int, specify a genome region in which variants will be ignored during imputation process.
-b, --exclude_file   file_path      # string, multiple genome regions user want to mask during imputation process.
-f, --MAF_cutoff     MAF_cutoff     # double, minimum MAF threshold for variants in genome reference panel, 0.01 in default.
-l, --lambda_value   lambda_value   # double, a constant value used to added with var-covariance matrix to gurantee the matrix is invertible, 0.1 in default. 
-w, --window_size    window_size    # int, Window size N, +-N/2 apart from molecular center pos, in base pair, 500000bp in default.
-d, --delta          delta_value    # double, is a hyper-parameter controls the declining rate of the effects of SNPs far from the trait.
```

### Running sample data
The sample of summary statistics are generated from eQTL summary data of CD4+ T cells from a single-cell eQTL study by Monique et. al, published in [Nature Genetics](https://www.nature.com/articles/s41588-018-0089-9) in 2018. The input eQTL summary file includes 24,846 variant-gene associations representing 100 genes and 5106 variants on chromosome 1. The sample of reference panel is generated from 1000G phase3 (EUR population) with variants having MAF>0.01. 198,337 associaitons including the inputs will be generated by xQTLImp. The whole process will take approximately 5~10 mins.

```bash
cd ./src ; make; cd .. # compliling under src folder
mkdir sample_output    # creat a new folder for output
# Running xQTLImp
./src/xQTLImp-dist                                  \
-m ./sample/sample_gene_annotation.txt         \
-x ./sample/sample_eQTL_summary.txt            \
-v ./sample/                                   \
-o ./sample_output/                            \
-t 1                                           \
-w 500000 \
-d 10
```

